src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Conversion.cs

```diff
@@ -406,9 +405,20 @@ private static bool IsFloatingPointExpressionOfUnknownPrecision(BoundExpression
                         var receiver = (!method.RequiresInstanceReceiver && !oldNodeOpt.IsExtensionMethod && !method.IsAbstract) ? _factory.Type(method.ContainingType) : mg.ReceiverOpt;
                         Debug.Assert(receiver is { });
                         _factory.Syntax = oldSyntax;
-                        return new BoundDelegateCreationExpression(syntax, argument: receiver, methodOpt: method,
-                                                                   isExtensionMethod: oldNodeOpt.IsExtensionMethod, type: rewrittenType);
+
+                        Debug.Assert(_factory.TopLevelMethod is { });
+                        if (DelegateCacheRewriter.CanRewrite(_factory.Compilation, _factory.TopLevelMethod, _inExpressionLambda, oldNodeOpt, method))
```

Comment:
>_inExpressionLambda [](http://example.com/codeflow?start=108&length=19)

I think we should check ```_inExpressionLambda``` explicitly prior to calling the method instead. In my opinion that is going to improve readability. Also, checking the value isn't really about whether we are able to cache the delegate, but rather about whether we even want to consider to cache it. The language version should be checked here as well, I think.

Reply:
Renamed to `DelegateCreationRewriter.AllowCaching` to reflect that we're also considering whether we should do it.

Reply:
> Renamed to DelegateCreationRewriter.AllowCaching to reflect that we're also considering whether we should do it.

I strongly suggest following the original suggestion because keeping the mentioned check explicitly here rather than burying them in a helper "is going to improve readability" in my opinion.

Reply:
OK, please check fef8b7a.